src/adapter-types/src/compaction.rs

```diff
@@ -7,25 +7,100 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
-// TODO: We can have only two consts here, instead of three, once there exists a `const` way to
-// convert between a `Timestamp` and a `Duration`, and unwrap a result in const contexts. Currently
-// unstable compiler features that would allow this are:
-// * `const_option`: https://github.com/rust-lang/rust/issues/67441
-// * `const_result`: https://github.com/rust-lang/rust/issues/82814
-// * `const_num_from_num`: https://github.com/rust-lang/rust/issues/87852
-// * `const_precise_live_drops`: https://github.com/rust-lang/rust/issues/73255
+use std::num::TryFromIntError;
+use std::time::Duration;
 
 use mz_repr::Timestamp;
-use std::time::Duration;
+use mz_storage_types::read_policy::ReadPolicy;
+use serde::Serialize;
+use timely::progress::frontier::MutableAntichain;
+use timely::progress::{Antichain, Timestamp as TimelyTimestamp};
 
 /// `DEFAULT_LOGICAL_COMPACTION_WINDOW`, in milliseconds.
 /// The default is set to a second to track the default timestamp frequency for sources.
 const DEFAULT_LOGICAL_COMPACTION_WINDOW_MILLIS: u64 = 1000;
 
-/// The default logical compaction window for new objects
-pub const DEFAULT_LOGICAL_COMPACTION_WINDOW: Duration =
-    Duration::from_millis(DEFAULT_LOGICAL_COMPACTION_WINDOW_MILLIS);
-
 /// `DEFAULT_LOGICAL_COMPACTION_WINDOW` as an `EpochMillis` timestamp
 pub const DEFAULT_LOGICAL_COMPACTION_WINDOW_TS: Timestamp =
     Timestamp::new(DEFAULT_LOGICAL_COMPACTION_WINDOW_MILLIS);
+
+/// The value to round all `since` frontiers to.
+/// We pick 1s somewhat arbitrarily, but matching historical practice.
+// TODO[btv] If we want to further reduce capability chatter, we can implement the design in
+// `20230322_metrics_since_granularity.md`, making it configurable.
+pub const SINCE_GRANULARITY: mz_repr::Timestamp = mz_repr::Timestamp::new(1000);
+
+// A common type (that is usable by the sql crate and also can implement various methods on types in
+// storage) to express compaction windows.
+#[derive(Clone, Copy, Debug, Eq, Ord, PartialEq, PartialOrd, Serialize)]
+pub enum CompactionWindow {
+    /// Unspecified by the user, use a system-provided default.
+    Default,
```

Comment:
I'd really prefer to not have this so it must be explicit every time. It's not 100% clear to me if everyone would expect `Default` instead of `DisableCompaction`.